File: linked_list/reorder_list.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class TestSolution(object):
    def reorderList(head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        # slow and fast iterators
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # group second half and cut in half
        second = slow.next
        prev = slow.next = None
        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp
            
        first, second = head, prev
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2

if __name__ == '__main__':
    l1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, None))))
    TestSolution.reorderList(l1)
        

File: linked_list/reorder_list.js
class ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

var reorderList = function (head) {
  // slow and fast?
  if (!head) return null;
  let slow = head;
  let fast = head.next;
  // get half of list
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  // slow is our half -> reverse
  let second = slow.next;
  slow.next = null;
  // reverse the half
  let reverse = second ? new ListNode(second.val) : null;
  while (second && second.next) {
    reverse = new ListNode(second.next.val, reverse);
    second = second.next;
  }

  let [prev, latter] = [head, reverse];
  while (latter) {
    let tmp1 = head.next;
    let tmp2 = latter.next;
    prev.next = latter;
    latter.next = tmp1;
    prev = tmp1;
    latter = tmp2;
  }
};
let x = ListNode(1, ListNode(2, ListNode(3, ListNode(4))));
reorderList(x);
